### PR TITLE
XRT Application Capture and Replay

### PR DESCRIPTION
**Pull request by stsoe**
#### Problem solved by the commit

Applications using XRT often involve complex, multi-layered software stacks (VAIML, PyTorch, TensorFlow) that make debugging, performance analysis, and regression testing difficult. There was no mechanism to capture the low-level XRT API interactions and replay them independently of the application stack. This commit introduces a capture and replay system that records XRT execution at the frame level (xrt::run::start() or xrt::runlist::execute() boundaries) and enables replay through a standalone executable.

#### How problem was solved, alternative solutions (if any) and why they were rejected

The solution uses a frame-based capture approach with buffer invalidation tracking. Key design decisions:

**Capture Strategy**: Uses polymorphic wrapper classes (run_impl_debug, runlist_impl_debug) to intercept XRT API calls with minimal performance impact when disabled. Macro `XRT_REPLAY_CAPTURE` with `XRT_UNLIKELY` branch hint ensures near-zero overhead when capture is disabled (capture_frames=0).

**Buffer Management**: Implements BO invalidation - buffer data is only dumped when xrt::bo::sync(XCL_BO_SYNC_BO_TO_DEVICE) is called. Run objects track which buffers are valid, avoiding redundant dumps when the same data is used across multiple frames.

**Artifact Deduplication**: Uses FNV-1a hash to deduplicate xclbins, ELF files, and buffer data, preventing duplicate storage of identical artifacts.

**Wait Synchronization**: Captures xrt::run::wait() and xrt::runlist::wait() calls and associates them with the current active frame. Replay executes waits in the same order, correctly handling asynchronous execution.

#### Risks (if any) associated the changes in the commit

1. **Performance when enabled**: Capture adds file I/O overhead proportional to buffer sync frequency. Measured impact is acceptable for debug/analysis workflows but should not be enabled in production.

2. **Thread safety**: The frames singleton uses std::mutex to protect shared state. All capture functions are thread-safe, but heavy contention under high-frequency API calls from multiple threads could cause slowdown.

3. **Memory growth**: Long-running captures with many unique ELF files will accumulate memory until process exit. Acceptable for typical debug sessions but could be problematic for extended captures.

4. **Breaking changes**: None. Capture is disabled by default and requires explicit xrt.ini configuration.

#### What has been tested and how, request additional testing if necessary

**Tested**:
- Basic capture/replay with single run per frame
- Runlist capture/replay with multiple runs per frame  
- Buffer invalidation with repeated xrt::bo::sync() calls
- ELF flow (programs) and xclbin flow
- Artifact deduplication (same xclbin/buffer used multiple times)
- Wait synchronization across multiple queued frames
- Stream seeking fix for elf_ctor (istream constructor)

**Recommended additional testing**:
- Stress test: 1000+ frames, 100+ runs per frame
- Large buffers (>1GB) to validate mmap efficiency
- Multi-threaded applications with concurrent API calls
- Edge cases: run started before set_arg, runlist with no runs
- Performance regression test with capture disabled (should be <1% overhead)
- Validation: compare output buffers from original vs replayed execution

#### Documentation impact (if any)

New documentation added in [src/runtime_src/core/common/runner/replay.md](src/runtime_src/core/common/runner/replay.md) covering:
- Quick start guide with xrt.ini configuration
- Replay executable usage and command-line options
- JSON schema specification
- Use cases (debugging, performance analysis, regression testing)
- Troubleshooting guide
- Advanced topics (artifact deduplication, ELF vs xclbin flow, module caching)

Configuration changes in xrt.ini:
```ini
[Runtime]
capture_frames=<num>          # Number of frames to capture (0=disabled, default)
capture_output_dir=<path>     # Output directory (default: ./)
New executable: xrt-replay for replaying captured execution.

### CODE REVIEW

**Comment by stsoe** [comment]
@sonals .  This is a monster PR, but I think safe since capture is non-intrusive and default disabled.   I need to do a lot more testing, but feel like getting these changes off my hand for now.   You may want to browse to the replay.md file in this PR, the md file is written by Claude.  The code was written by me, but reviewed by Claude.

File: src/runtime_src/core/common/api/xrt_bo.cpp [hunk 1]
```diff
@@ -482,6 +495,8 @@ class bo_impl
     // if (get_flags() != bo::flags::host_only)
     handle->sync(static_cast<xrt_core::buffer_handle::direction>(dir), sz, offset);
     m_usage_logger->log_buffer_sync(device->get_device_id(), device.get_hwctx_handle(), sz, dir);
+
+    XRT_REPLAY_CAPTURE(bo_sync, this, static_cast<int>(dir));
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/api/xrt_bo.cpp:499]: On purpose.

File: src/runtime_src/core/common/runner/capture.cpp [hunk 2]
```diff
@@ -0,0 +1,1190 @@
...
+      capture_frame_start_data(ft, repo);
+    }
+
+    const void*
+    get_handle() const
+    {
+      return std::visit([](const auto& v) -> const void* {
+        using T = std::decay_t<decltype(v)>;
+        if constexpr (std::is_same_v<T, run*>)
+          return v->get_handle();
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/runner/capture.cpp]: Required for compilation to resolve.

File: src/runtime_src/core/common/runner/capture.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{
+  static auto frames = xrt_core::config::get_capture_frames();
+  if (!frames)
+    return false;
+
+  return num_frames() < frames;
+}
+
+// No-op when disabled, actual capture when enabled
+#define XRT_REPLAY_CAPTURE(fn, ...) \
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/runner/capture.h:25]: On purpose

File: src/runtime_src/core/common/runner/capture.cpp [hunk 4]
```diff
@@ -0,0 +1,1199 @@
...
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/runner/capture.cpp]: Necessary to make compiler happy

File: src/runtime_src/core/common/runner/detail/capture_artifacts.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    auto hash = 0xcbf29ce484222325ULL;
+    for (unsigned char c : data) {
+      hash ^= c;
+      hash *= 0x100000001b3ULL;
+    }
+    return hash;
+  }
+
+public:
+  artifacts(std::filesystem::path dir)
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/runner/detail/capture_artifacts.h:61]: Fixed

File: src/runtime_src/core/common/runner/capture.cpp [hunk 6]
```diff
@@ -0,0 +1,1190 @@
...
+    
+    static uint64_t
+    get_uid()
+    {
+      static std::atomic<uint64_t> id{0};
+      return id++;
+    }
+
+  public:
+    bo(xrt::bo bo)
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/runner/capture.cpp:125]: Fixed

File: src/runtime_src/core/common/runner/detail/module_cache.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  if (auto it = s_elf2mod.find(elf); it != s_elf2mod.end())
+    return (*it).second;
+
+  xrt::module mod{elf};
+  s_elf2mod.emplace(elf, mod);
+  return mod;
+}
+
+static xrt::module
+get(const std::string& path, const repo_type& repo)
```
[the remark this answers is not in the document]
Comment (stsoe) [reply src/runtime_src/core/common/runner/detail/module_cache.h:35]: Need to be declared inline
